server/uno_test.go

```diff
@@ -8,63 +8,118 @@ import (
+	game, player := setupGameWithPlayer(database)
+
+	drawCardHelper(game, player)
+
+	assert.Equal(t, len(player.Cards), 1)
+}
+
 func TestDrawCard(t *testing.T) {
 
-	// database, _ := db.GetDb()
+	database, _ := db.GetDb()
+
+	game, player := setupGameWithPlayer(database)
+
+	// Test Drawing a card with a full deck and real player
+	game, err := drawCard(game.ID, player.ID)
+	game, _ = database.LookupGameByID(game.ID)
+	player = &game.Players[game.CurrentPlayer]
+
+	// Assert that no error occured, the player has a new card and the draw pile
+	// has one less card
+	assert.Nil(t, err, "Failed to draw card.")
+	assert.Equal(t, 1, len(player.Cards))
+	assert.Equal(t, 107, len(game.DrawPile))
+
+	// Move all cards into the discard pile, Empty out the draw pile completely,
+	// and test drawing a card. It should resuffle leaving one card on the discard pile
+	game.DiscardPile = append(game.DiscardPile, game.DrawPile...)
+	game.DrawPile = game.DrawPile[:0]
+	lastCard := game.DiscardPile[len(game.DiscardPile)-1]
+
+	database.SaveGame(*game)
+
+	game, err = drawCard(game.ID, player.ID)
+	player = &game.Players[game.CurrentPlayer]
+
+	//Assert no error, player has 2 cards from both draw tests,
+	// draw is missing three from two draws and one in discard
+	// discard has one last remaining card.
+	// Assert last card in discard is actually to proper last card
+	assert.Nil(t, err, "Failed to draw card.")
+	assert.Equal(t, 2, len(player.Cards))
+	assert.Equal(t, 105, len(game.DrawPile))
+	assert.Equal(t, 1, len(game.DiscardPile))
+	assert.Equal(t, lastCard.Color, game.DiscardPile[0].Color)
+	assert.Equal(t, lastCard.Value, game.DiscardPile[0].Value)
+
+	// Empty out both discard and draw piles, it should for now put a new deck on the draw pile.
+	game.DrawPile = game.DrawPile[:0]
+	game.DiscardPile = game.DiscardPile[:1]
+	lastCard = game.DiscardPile[len(game.DiscardPile)-1]
+
+	database.SaveGame(*game)
+
+	game, err = drawCard(game.ID, player.ID)
+	player = &game.Players[game.CurrentPlayer]
+
+	// Assert no errors, assert player now has 3 cards
+	// assert new draw pile with one missing
+	// assert discard still has one card
+	// Assert last card in discard is actually to proper last card
+	assert.Nil(t, err, "Failed to draw card.")
+	assert.Equal(t, 3, len(player.Cards))
+	assert.Equal(t, 107, len(game.DrawPile))
+	assert.Equal(t, 1, len(game.DiscardPile))
+	assert.Equal(t, lastCard.Color, game.DiscardPile[0].Color)
+	assert.Equal(t, lastCard.Value, game.DiscardPile[0].Value)
+
+	// Create a player to simulate one playing but its not their turn and attempt to play
+	otherPlayer := model.Player{ID: " id 2 ", Name: "Name 2", Cards: []model.Card{}}
 
-	// game, player := setupGameWithPlayer(database)
+	_, err = drawCard(game.ID, otherPlayer.ID)
 
-	// game, err := drawCard(game.ID, player.ID)
+	// Assert that we got an error from the draw card.
+	assert.NotNil(t, err, "Player played out of turn. Please make sure only the player whoes turn it is can play.")
```

Comment:
What's the third argument here? Does `assert.NotNil` just check that `err != nil` and if so what's the use of the string? I don't think it matches what the error will be

Reply:
The String is what you get in the console if the test case fails. Just before this i simulate a player trying to draw a card out of turn. I create a bogus player with a bogus id and pass them into Draw Card. We should get an error message so i check that the error exists. It should not allow them to draw a card. Only the players who's turn it is should be able to draw a card.  I should probably assert a couple more items around both the bogus player and the game. Also it should say "Player 'drew' a card out of turn".

Reply:
I should check that the message from the error is also the one i expect. There are two other places in drawCard that an error can be returned.